### ISSUE TITLE
Architecture Proposal

### ISSUE THREAD

**Issue by coriolinus** [exercism/automated-tests#6]
## Docker Architecture

- Exercism publishes a docker container which is pretty much nothing but Alpine, `git`, and the `exercism` tool
- Each track team builds and publishes a docker container derived from the exercism general-purpose container. Each such container adheres to the following convention:
    - it has an environment variable `SOLUTION_ID` which the `exercism` tool can use to download the student's solution
    - it has an environment variable `GIT_SHA` identifying the track's git SHA at the time the student began the exercise
    - It has an entrypoint which downloads the test suite, the student's solution, runs one against the other, and prints a JSON report of the results to stdout
- whenever a student submits a solution, exercism infrastructure kicks off the test suite with `docker run --rm -e SOLUTION_ID=... -e GIT_SHA=... exercism/test-track-${TRACK}:latest` and watches its stdout. 
- docker will take care of monitoring what the latest version of each track's test runner is, and updating itself on demand

Note: the test runner must attempt to run all optional tests. However, these may be failed without penalty.

## JSON Report Format

Every 10 seconds or so while the test suite is running, it should emit an object like

```json
{"status":"running"}
```

This ensures we don't have to wait too long for timed-out or hung processes: if no JSON message shows up in 20 seconds, we can kill the container and report that it hung due to an internal error.

We should also kill the container after 10 minutes: no test suite should last that long. If we got there, most likely the runner is fine, but a student's solution is inefficient or malicious.

Once there are results, the test runner will emit an object like this:

```json
{
  "status":"complete",
  "tests": [
    {
      "name": "test_foo_did_bar",
      "result": "pass"
    },
    {
      "name": "test_foo_did_bat",
      "result": "fail",
      "stdout": "..."
    },
    {
      "name": "test_foo_did_baz",
      "result": "fail",
      "stdout": "...",
      "optional": "this test is feature-gated behind the 'baz' feature"
    }
  ]
}
```

**Comment by mawis** [comment]
> Each track team builds and publishes a docker container derived from the exercism general-purpose container.

Any special reason for this definition?

**Comment by coriolinus** [comment]
Just trying to reduce work. It could be just as efficient computationally if each track had to build their own container from scratch, but that wouldn't be an efficient use of programmer time. It's simpler to just publish the common base container and derive from that, so that each track just needs to add to it the exact dependencies it needs to run an arbitrary test suite.

**Comment by mawis** [comment]
I totally agree, that work shouldn't be duplicated. I just wouldn't make it a requirement. With some languages it could just be easier not using the common base image because they have different common practices.

For some time already (without yet proposing it), I thought it would be nice to have a [Nix Expression Language](https://nixos.wiki/wiki/Nix_Expression_Language) track. In the world of Nix docker containers are often build with their own tools, that don't need base images at all.

This is explained more in [this article by Luca Bruno](http://lethalman.blogspot.com/2016/04/cheap-docker-images-with-nix_15.html) in results in small containers comparable to Alpine as well.

**Comment by coriolinus** [comment]
I see. I'm fine amending to "track test containers may extend the base
image", which should give less a sense that it is mandatory.

**Comment by mawis** [comment]
I could also live with “should”.

**Comment by SaschaMann** [comment]
> Once there are results, the test runner will emit an object like this:

I would strongly prefer a standardized output format, like junit test reports. It seems pointless to reinvent the wheel when there already are formats to store test results that are supported in dozens of languages.

---

If you want to use a non-standard own format, there should be support for grouping tests together. This is how problem-specs handles it and it's also quite common in unit testing frameworks.

What will the `name`s be used for? It's not always canonical or even possible to label individual tests so I'm wondering whether this could also be an optional field or if the names will be required to display them on the website.

**Comment by iHiD** [comment]
Thanks for kicking off the discussion!

A few quick thoughts:
1) I imagine we're going to want to keep the rules around Docker images and the interface broadly in line with what we have for automated analysers. Details of those can be found here: [Docker](https://github.com/exercism/automated-analysis/blob/master/docs/docker.md) and [Interface](https://github.com/exercism/automated-analysis/blob/master/docs/docker.md). Obviously the actual output etc will change, but I imagine the rest can be quite similar.
2) We currently have a 10s timeout and resource-limit on analyzers. Again, I imagine we'll want to keep these pretty similar as we've based them on scale and (financial) resources. **If your test suite takes longer than 10s to run for an exercise, please let me know**
3) I really like the idea of a JSON format. Other people I've chatted with privately really agree think that we should use something like junit. I suspect this is going to be a key point of contention and I might host a Slack discussion on it, where we can flesh out the pros/cons.
4) The tests **will** be displayed as individual results in the website, so we will want things like the name/status etc to be broken out as above.
5) I'd like tests to fail after the first failure (to use minimal resources), so have a think about how your language's test runner can achieve that.

**Comment by coriolinus** [comment]
- "Interface" link is broken, above; it points to "Docker"
- It's fairly common in Rust for compilation alone to take more than 10s. More on that later.
- Junit format is possible if really necessary, but I'm not really a fan of XML formats; they're just too verbose. 
- A point in favor of writing our own format: we can include data about optional tests in a straightforward way. I have no idea if Junit format can express that information within its schema. 

With regard to total execution time, 10 seconds is too low for Rust. Our exercises are usually pretty quick to compile, unless students have included external dependencies. At that point, it can take most of 10 seconds to download and compile all the dependencies. Several exercises and feature-gated tests are best solved with certain external dependencies. We also have individual several test cases which take longer than 10s to run in debug mode. 

I [wrote a quick script](https://gist.github.com/coriolinus/7fe8d95961f698160a0a25654b518f42) to see what the worst offenders actually are on the current `master` branch. This is on my machine, which is faster than most cheap cloud instances you'll find right now. This uses the example solutions for each exercise, which are typically fairly well optimized. Student implementations may perform worse. These are the results:

```
31.44 alphametics
28.01 palindrome-products
11.40 scale-generator
11.22 grep
8.21 pig-latin
6.68 decimal
4.66 gigasecond
4.14 crypto-square
3.79 diffie-hellman
3.61 simple-cipher
```

**Comment by iHiD** [comment]
@coriolinus Thanks for digging into the details. We'll have to work out what's possible for us from a cost POV. If we. We currently have (on average) one submission per minute. Presuming a 10x increase in the next few months (which is realistic) we need to be able to keep up with one submission every 5-6s. 

Are external dependencies in Rust something that are relied on to solve the exercises? I know in many other languages we limit people to just the core + std libs.

Worst case we can the auto-running of tests in some exercises to not be supported, but it would be nice if we can avoid that.

(Link to interface is updated)

**Comment by mawis** [comment]
@iHiD All Rust core exercises can be solved without using external dependencies. I think there is just one core exercise, that at least encourages the student to use an additional library ("crate") by having an optional test, that nearly requires to use this additional crate. There is even a comment generated by auto-mentor that encourages the user to implement that test as well. (I don't know all of the side exercises, so I don't know how much of them use external crates.)

Sadly it's very common in Rust development, that implementations often fetch a whole lot of crates. I have the same problem at work, when I'm optimizing build pipelines for embedded software/firmware and wonder why a product, that does no text handling at all loads crates for handling Japanese text or things like that. Well it's just a dependency of a dependency of a dependency of some library that in some special cases might have to handle eastern Asian text. - It never happens in my product, but increases my build times by minutes.

**Comment by mawis** [comment]
Oh right ... the second exercise after hello world (Gigasecond) uses a crate as well. But according to coriolinus timings it builds fast enough.

**Comment by mawis** [comment]
I checked it for the PureScript track. All exercises have the same (by default already added) dependencies there. Compiling just these on my machine also takes already exactly 10 seconds.

About the same time for exercises I did in Kotlin or Java where it takes time to start the build tool gradle.

10 seconds might become a problem for compiled languages ...

**Comment by neenjaw** [comment]
Question about the git SHA, when a solution is requested, are the test cases not returned with it?

What would the role of the git SHA be?

**Comment by NobbZ** [comment]
@coriolinus:

> Exercism publishes a docker container which is pretty much nothing but Alpine, `git`, and the `exercism` tool

What do we need `git` and `exercism` for?

From the experiences with the analyzer I had assumed that we provide a wrapper script that gets called with the path to the solution under test.

That way I neither need the CLI, nor `git` as everything is already there.

This basically brings us back to an empty `alpine:…`, I prefer to choose my own then, as there are known problems with erlang and alpine 3.10.

> ** it has an environment variable SOLUTION_ID which the exercism tool can use to download the student's solution
> ** it has an environment variable GIT_SHA identifying the track's git SHA at the time the student began the exercise

In my opinion this overcomplicates things, just give us a volume that already contains the students solution, and start the wrapper with a single argument as explained above.

Perhaps add the exercises name, as some testrunners might need it to infer the actual command to run the tests, as its name changes with the exercise (eg. ruby, bash, python…).

> docker will take care of monitoring what the latest version of each track's test runner is, and updating itself on demand

You mean as in publishing those images to the docker hub? Or to the GitHub registry or somewhere else? Especially if it is the docker hub, who shall be the uploader? Do you really want to hand out credentials for a single account to 50+ people? Create some CI that pushes them?

Those questions are unanswered…

I prefer the review and publish process thats currently in use for the analyzer project as well. At least during the starting phase.

> Every 10 seconds or so while the test suite is running, it should emit an object like

This can be hard in a lot of languages.

> Once there are results, the test runner will emit an object like this:
> [large JSON dump]

I'm not sure if I can hook into erlangs test runner to actually get the info to build such a JSON. Or any other kind of format that is not plain what I got from stdout/err of the testrunner. And I really do not want to parse this output, as it might change with the version of the runner.

--------------

@SaschaMann:

> I would strongly prefer an existing, standardized output format, like junit test reports.

Basically the same applies, I'm not sure if I am actually able to get all the necessary information from the testrunner, but in general, If I had to implement a formatter, I'd really prefer some pre-existing and well known format, as perhaps someone else might already have written such a formatter I just need to use…

**Comment by coriolinus** [comment]
> What would the role of the git SHA be?

> What do we need git and exercism for?

Just different initial assumptions about how the distribution of work should go. I haven't worked on the analyzer, so I was inventing the architecture from scratch. As I imagined it, the wrapper script didn't need to do any work at all to prepare the field; the docker container would download both the appropriate test suite and the student's solution, given those two env variables. 

Between my proposed architecture and the one @iHiD sketched, based on the analyzer's, I don't have a preference. It appears that there's strong preference here for the existing architecture, which is fine with me.

> > Every 10 seconds or so while the test suite is running, it should emit an object like
> This can be hard in a lot of languages.

If the total time before killing a container is less than 30s or so, we don't really need this feature. It's a useful status indicator for longer-running jobs, but these may not be. 

However, this is where I'd expect a wrapper script within the docker container: we'd write some process which ran the test suite as a child process, emitted heartbeat messages, and parsed/rewrote the test suite's output into our desired report format. Keep in mind, there is no reason at all that this wrapper must be written in the same language as the one being tested. I'm most familiar with Rust, Python, and Go; writing such a wrapper script would be very simple in any of those. I'm pretty confident that we could also get this behavior in straight bash. 

I agree that it's hard to imagine how to do this in something like Prolog, but as long as we assume that maintainers of languages where this is hard are also familiar with at least one language where it is not hard, then it's not unreasonable to design the system to require a wrapping script within the container.

> I'm not sure if I can hook into erlangs test runner to actually get the info to build such a JSON... And I really do not want to parse this output,

That's a reasonable objection. It seems to me that fundamentally there are two options: we can present the test results as the straight output from the language's test runner (possibly collecting multiple runs of data, in the case of optional features), or we can parse the data into a common format and let Exercism's UI team decide how to present the result.

The first option is easy to implement. The second option is a lot more user-friendly. I have a mild preference for the second, but I'm sympathetic to your position.

**Comment by NobbZ** [comment]
> However, this is where I'd expect a wrapper script within the docker container: we'd write some process which ran the test suite as a child process, emitted heartbeat messages, 

Such a wrapper emitting the heartbeat, independent of the actual status of the child is meaningless.

**Comment by SaschaMann** [comment]
> Junit format is possible if really necessary, but I'm not really a fan of XML formats; they're just too verbose.

I'm not a huge fan of the format itself either but it's so widely used and supported that this would save a huge amount of work compared to writing custom test output formatters.

> A point in favor of writing our own format: we can include data about optional tests in a straightforward way. I have no idea if Junit format can express that information within its schema.

The format itself supports it, it allows different failure types, but I'm not sure if all test report generators have an easy way to add this information.

> The tests will be displayed as individual results in the website, so we will want things like the name/status etc to be broken out as above.

Some exercises on the Julia track have a large amount of individual tests (`rational-numbers` has nearly 95000, `robot-name` nearly 250), so I hope there will be some way to group them.

**Comment by coriolinus** [comment]
> Such a wrapper emitting the heartbeat, independent of the actual status of the child is meaningless.

Disagree: 

- it's a reliable indicator that whatever's going on in that container, it didn't just die. 
- who says it's independent of the actual status of the child? It's not hard to watch to see if a child process is still alive or dead; if it's alive, it's still processing. If it's dead, you can return an actual error.

**Comment by coriolinus** [comment]
>> Junit format is possible if really necessary, but I'm not really a fan of XML formats; they're just too verbose.
> 
> I'm not a huge fan of the format itself either but it's so widely used and supported that this would save a huge amount of work compared to writing custom test output formatters.

That's true or not on a track-specific basis. For Rust, you can natively tell the test suite to emit a JSON report; from there, it's mainly a matter of applying some `jq` filters (or equivalent) and ensuring that optional features are included. Conversely, there isn't any built-in Junit support.

I suspect that the only fair choice between Junit and JSON will come from actually tabulating which tracks have support for each of those, and then taking the majority vote.

**Comment by NobbZ** [comment]
> it's a reliable indicator that whatever's going on in that container, it didn't just die.

If the worker just dies, then docker knows… If the worker infiniloops, no one knows and we will time out eventually. The heartbeat would not change that.

I'd just omit the heartbeat and consider any output to stdout as "healthy" similar to how travis handles it. Jobs will timeout after they haven't printed anything for 10 minutes.

> who says it's independent of the actual status of the child? It's not hard to watch to see if a child process is still alive or dead; if it's alive, it's still processing. If it's dead, you can return an actual error. It's a loose coupling, sure, but that's a very different thing than being independent.

As I said, docker knows about crashes, just let it handle them and provide full logs to the developers. While simply writing the actual response to a file into a well known place, similar to how its done with the analyzers.

**Comment by coriolinus** [comment]
Hmm. I'm a bit skeptical that Docker knows about crashes by default and
automatically does the right thing; I've had the experience of containers
just stopping with no output. However, it's entirely possible that this was
the result of misconfiguration. I'm willing to posit that getting Docker to
print a crash log (as applicable) is possible, and that if it is possible,
that's a better solution than a heartbeat.

**Comment by mawis** [comment]
NobbZ:
> If the worker just dies, then docker knows… If the worker infiniloops, no one knows and we will time out eventually. The heartbeat would not change that.

coriolinus:
> I'm a bit skeptical that Docker knows about crashes by default and
automatically does the right thing

In case of a real crash (program exiting due to an exception or a signal) docker should know. If you started the container with `docker run` then the call returns outside the container as well and everything should be fine. If you started the container with `docker start` what happens depends on further configuration (restarting, exiting, …)

In case of an infinite loop or some threads blocking each other, docker doesn't detect that but I think 
NobbZ didn't claim that. A health check could continue sending pings in that case anyway as well. But that's what the execution limit is for, isn't it?

**Comment by NobbZ** [comment]
My claim was that infinite loops wouldn't be recognized either way.

No docker, nor wrapped heartbeats, nor internal heartbeats would be able to detect those…

Well, perhaps an internally implemented heartbeat that would block the heartbeat "thread" because the looping "thread" doesn't hit a boundary that would make the scheduler rescheduling the "threads"… But I think this is an edge case thats rarely hit and does not justify the heartbeats additional complexity.

**Comment by coriolinus** [comment]
Agree that a heartbeat wouldn't detect an infinite loop: that's what the
max time bound is for. The point of a heartbeat is to detect a
hang/deadlock which has prevented the test runner from emitting any output
for longer than expected.

**Comment by iHiD** [comment]
Helpful discussion. Thank you all. 

We've worked through quite a lot of this with the analyzers, so I'll discuss it with @ccare (who has done the work on that) and can fill in some more details of how things work there. I will then take some time in the next few days to write a proposal up and then we can discuss/debate that as a tangible thing. I think the key will be understanding whether there are any limitations/standards in different languages (e.g. the coverage of junit format vs writing JSON) so I will try to list those as discussion points on that issue.

I suggest leaving this discussion until then :)

**Comment by iHiD** [comment]
I'm closing this in favour of 
https://github.com/exercism/automated-tests/pull/13
https://github.com/exercism/automated-tests/pull/14

Note that I've decided that I want a custom JSON format as the output, because I want it tightly-coupled to the application's needs. However, we'll write a JUnit->JSON script that you can run as the end of the Docker pipeline, so you can use JUnit if you want to. We can also write similar scripts for other reporters as necessary.
